src/ui/Frame.cpp

```diff
@@ -0,0 +1,1164 @@
+						actualSize.x = std::min(std::max(0, actualSize.x), std::max(0, actualSize.w - size.w));
+						mousestatus[SDL_BUTTON_LEFT] = 0;
+					}
+					usable = result.usable = false;
+					ticks = -1; // hack to fix sliders in drop downs
+				}
+			}
+		}
+
+		// vertical slider
+		if (actualSize.h > size.h) {
+			// rail
+			SDL_Rect sliderRect;
+			sliderRect.x = _size.x + _size.w;
+			sliderRect.y = _size.y;
+			sliderRect.w = sliderSize;
+			sliderRect.h = _size.h;
+
+			// handle
+			float winFactor = ((float)_size.h / (float)actualSize.h);
+			int handleSize = std::max((int)(size.h * winFactor), sliderSize);
+			int sliderPos = winFactor * actualSize.y;
+			SDL_Rect handleRect;
+			handleRect.x = _size.x + _size.w;
+			handleRect.y = _size.y + sliderPos;
+			handleRect.w = sliderSize;
+			handleRect.h = handleSize;
+
+			// click & drag
+			if (draggingVSlider) {
+				if (!mousestatus[SDL_BUTTON_LEFT]) {
+					draggingVSlider = false;
+				} else {
+					float winFactor = ((float)_size.h / (float)actualSize.h);
+					actualSize.y = (mousey - omousey) / winFactor + oldSliderY;
+					actualSize.y = std::min(std::max(0, actualSize.y), std::max(0, actualSize.h - size.h));
+				}
+				usable = result.usable = false;
+				ticks = -1; // hack to fix sliders in drop downs
+			} else {
+				if (rectContainsPoint(handleRect, omousex, omousey)) {
+					if (mousestatus[SDL_BUTTON_LEFT]) {
+						draggingVSlider = true;
+						oldSliderY = actualSize.y;
+					}
+					usable = result.usable = false;
+					ticks = -1; // hack to fix sliders in drop downs
+				} else if (rectContainsPoint(sliderRect, omousex, omousey)) {
+					if (mousestatus[SDL_BUTTON_LEFT]) {
+						actualSize.y += omousey < handleRect.y ? -std::min(entrySize * 4, size.h) : std::min(entrySize * 4, size.h);
+						actualSize.y = std::min(std::max(0, actualSize.y), std::max(0, actualSize.h - size.h));
+						mousestatus[SDL_BUTTON_LEFT] = 0;
+					}
+					usable = result.usable = false;
+					ticks = -1; // hack to fix sliders in drop downs
+				}
+			}
+		}
+	}
+
+	// process buttons
+	{
+		auto prev = buttons.rbegin();
+		for (auto it = buttons.rbegin(); it != buttons.rend(); it = prev) {
+			Button* button = *it;
+			prev = std::next(it);
+
+			if (!destWidget) {
+				destWidget = button->handleInput();
+			}
+
+			Button::result_t buttonResult = button->process(_size, actualSize, usable);
+			if (usable && buttonResult.highlighted) {
+				result.highlightTime = buttonResult.highlightTime;
+				result.tooltip = buttonResult.tooltip;
+				if (buttonResult.clicked) {
+					button->activate();
+				}
+				result.usable = usable = false;
+			}
+
+			if (destWidget && button->isSelected()) {
+				button->deselect();
+			}
+		}
+	}
+
+	// process (widget) sliders
+	{
+		auto prev = sliders.rbegin();
+		for (auto it = sliders.rbegin(); it != sliders.rend(); it = prev) {
+			Slider* slider = *it;
+			prev = std::next(it);
```

Comment:
So is this really a previous or is it more of a next?

Reply:
std::next() for a reverse iterator is "previous" for a normal iterator :)

Reply:
Is this a reverse iterator? I see prev and iterator both start at rbegin, and prev seems to move on to the next item after it?

Reply:
rbegin() returns a reverse iterator.

Reply:
Also, I don't see `prev` being used anywhere. Why not do `++it` directly?

Reply:
> rbegin() returns a reverse iterator.

Aha. I completely glossed over the r, oops.

Reply:
I don't embed it = std::next(it) in the for loop because the iterator could be invalid after the process runs (for example, a button that deletes itself). the next position is therefore saved so the increment is valid

Reply:
actually i'm not sure how well tested that is since the next position could just as well be deleted, not to mention deletion happens in a later pass, so it might not be necessary

Reply:
Got it.